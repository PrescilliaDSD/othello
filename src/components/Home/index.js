import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

const Home = ({ setLoggedIn, gameData, setGameData, message, setMessage }) => {
  const [gamesList, setGamesList] = useState([])

  const logout = () => {
    setLoggedIn(false)
    sessionStorage.setItem('pseudo', '')
    sessionStorage.setItem('password', '')
    setMessage({
      text: '',
      status: ''
    })
  }

  const createGame = () => {
    const user = {
      pseudo: sessionStorage.getItem('pseudo')
    }
    fetch('http://localhost:8888/createGame', {
      method: 'POST',
      headers: {
        'Content-type': 'application/json'
      },
      body: JSON.stringify(user)
    })
      .then((res) => res.json())
      .then((res) => {
        setGameData(res.game)
        setMessage({
          text: 'Votre partie a bien été créée. Pour y accéder, cliquez sur le bouton "Accéder à la partie créée".',
          status: 'success'
        })
        setGamesList([...gamesList, res.game])
      })
  }

  useEffect(() => {
    fetch('http://localhost:8888/game', {
      method: 'GET',
      headers: {
        'Content-type': 'application/json'
      }
    })
      .then((res) => res.json())
      .then((res) => setGamesList([...res.data]))
  }, [])

  return (
    <main className="home">
      <h1>Jeu Othello</h1>
      <p className={`message message-${message.status}`}>{message.text}</p>
      <p className="what-to-do">Que voulez-vous faire ?</p>
      <ul className="home__actions">
        {
          gameData._id &&
          <li><Link className="button" to={`/game/${gameData._id}?pseudo=${sessionStorage.getItem('pseudo')}`}>Accéder à la partie créée</Link></li>
        }
        {
          !gameData._id &&
          <li><button onClick={createGame}>Créer une partie</button></li>
        }
        <li><button onClick={logout}>Me déconnecter</button></li>
      </ul>
      <p>Parties en cours : </p>
      <ul className="games__list">
        {
          gamesList.length > 0 && gamesList.map((game) => {
            if (game.status === 'wait for player 2') {
              return (
                <li key={game._id} className="games__list__item"><Link to={`/game/${game._id}?pseudo=${sessionStorage.getItem('pseudo')}&pseudo2=${sessionStorage.getItem('pseudo')}${2}`}>Partie {game._id}</Link></li>
              )
            }
            return <li key={game._id} className="games__list__item"></li>
          })
        }
      </ul>
    </main>

  )
}

Home.propTypes = {
  setLoggedIn: PropTypes.func.isRequired,
  gameData: PropTypes.object.isRequired,
  setGameData: PropTypes.func.isRequired,
  message: PropTypes.object.isRequired,
  setMessage: PropTypes.func.isRequired
}

export default Home
