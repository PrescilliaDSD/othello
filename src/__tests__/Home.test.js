import React from 'react'
import renderer from 'react-test-renderer'
import { BrowserRouter as Router } from 'react-router-dom'

import Home from '../components/Home'

describe('Root component', () => {
  it('should render correctly', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render a main tag', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree.type).toBe('main')
  })

  it('should render 6 children in the main tag', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree.children).toHaveLength(6)
  })

  it('should render h1, p, p, p, ul and ul in the main tag', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree.children[0].type).toBe('h1')
    expect(tree.children[1].type).toBe('p')
    expect(tree.children[2].type).toBe('p')
    expect(tree.children[3].type).toBe('ul')
    expect(tree.children[4].type).toBe('p')
    expect(tree.children[5].type).toBe('ul')
  })

  it('should render text "Jeu Othello" in the h1 tag', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree.children[0].children[0]).toBe('Jeu Othello')
  })

  it('should render a home actions list of two items', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree.children[3].children).toHaveLength(2)
  })
  it('should render two buttons "Créer une partie" & "Me déconnecter" in the home actions list if gameData is empty', () => {
    const component = renderer
      .create(<Home setLoggedIn={() => {}} gameData={{}} setGameData={() => {}} message={{}} setMessage={() => {}} />)
    const tree = component.toJSON()
    expect(tree.children[3].children[0].children[0].children[0]).toBe('Créer une partie')
    expect(tree.children[3].children[1].children[0].children[0]).toBe('Me déconnecter')
  })
  it('should render link "Accéder à la partie créée" & button "Me déconnecter" in the home actions list if gameData is not empty', () => {
    const component = renderer
      .create(<Router><Home setLoggedIn={() => {}} gameData={{ _id: '123' }} setGameData={() => {}} message={{}} setMessage={() => {}} /></Router>)
    const tree = component.toJSON()
    expect(tree.children[3].children[0].children[0].children[0]).toBe('Accéder à la partie créée')
    expect(tree.children[3].children[1].children[0].children[0]).toBe('Me déconnecter')
  })
})
