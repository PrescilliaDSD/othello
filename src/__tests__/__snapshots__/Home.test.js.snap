// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root component should render correctly 1`] = `
<main
  className="home"
>
  <h1>
    Jeu Othello
  </h1>
  <p
    className="message message-undefined"
  />
  <p
    className="what-to-do"
  >
    Que voulez-vous faire ?
  </p>
  <ul
    className="home__actions"
  >
    <li>
      <button
        onClick={[Function]}
      >
        Créer une partie
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
      >
        Me déconnecter
      </button>
    </li>
  </ul>
  <p>
    Parties en cours : 
  </p>
  <ul
    className="games__list"
  />
</main>
`;
